File: LeetCode/380-insert-delete-getrandom-o1.py
###############################################################################################
# easy
###########
# 时间复杂度：O(1)
# 空间复杂度：O(n)
###############################################################################################
class RandomizedSet:

    def __init__(self):
        self.hash, self.list = {}, []

    def insert(self, val: int) -> bool:
        if val not in self.hash:
            self.list.append(val)
            self.hash[val] = len(self.list) - 1
            return True
        return False

    def remove(self, val: int) -> bool:
        if val not in self.hash:
            return False
        lo, last = self.hash[val], self.list[-1]
        self.list[lo] = last
        self.hash[last] = lo
        self.list.pop()
        self.hash.pop(val)
        return True

    def getRandom(self) -> int:
        return self.list[random.randint(0, len(self.list)-1)]

File: LeetCode/380-insert-delete-getrandom-o1.cpp
///////////////////////////////////////////////////////////////////////////////////////////////
// just use rand() function in c++ and map for quick search and vector for random pick
////////////////
// Time Complexity: O(1)
// Space Complexity: O(n)
///////////////////////////////////////////////////////////////////////////////////////////////
class RandomizedSet {
public:
    vector<int> nums;
    unordered_map<int, int> m;
    
    RandomizedSet() {
        
    }
    
    bool insert(int val) {
        // already exist
        if (m.find(val) != m.end()) return false;
        
        nums.push_back(val);
        // mark the location in vector for delete
        m[val] = nums.size() - 1;
        
        return true;
    }
    
    bool remove(int val) {
        // not exist
        if (m.find(val) == m.end()) return false;
        
        // save the last number to the position of the removed one
        int last = nums.back();
        m[last] = m[val];
        nums[m[val]] = last;
        // pop the last one
        nums.pop_back();
        // erase from map
        m.erase(val);
        
        return true;
    }
    
    int getRandom() {
        return nums[rand() % nums.size()];
    }
};